### PR TITLE
1126 - Add mapper support to HL7 ingest

### PR DESCRIPTION
**Pull request by carlosfelix2**
This PR adds support to run mappers as part of the HL7 data ingest process.

How it works?
1. The HL7 ingest first gets all the values from the HL7 data
2. Then the mappers are run, overwriting any HL7 data obtained for the field
3. If the field value is empty then the default value is used

Testing:
Check the HL7 messages in src\test\resources\test_data_files\Hl7_ORU-R01 vs the CSV and verify the data is decided as expected.

## Changes
- Minor cleanups of lint warnings
- Added support to call the mappers
- Added a none() mapper to overwrite the use of mappers in the parent schema - I could not use the standard YAML values to null the mapper.
- New schema specific to HL7 ingest as the covid-19 schema is more geared towards CSV - this schema is a copy of the covid-19 schema with only the relevant HL7 configuration.
- Removed HL7 ingest specific configuration from covid-19 schema
- Unse alternate HL7 fields (hl7OutputFields) to look for values
- Updated ObservationTests to not be dependent on schema column order

### Testing
- [x] Tested locally?
- [x] Ran `quickTest all`?
- [x] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Security
- [x] Did you check for sensitive data, and remove any? 
- [x] Does logging contain sensitive data?  
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

### Process
- [x] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
- #1126 

## To Be Done
- Need to look at all the fields and identify any further decoding needed.
- I am at a good point that I want to split the ingest and generation portions of the HL7 serializer class into two separate classes.
- There are fields that gather funky values, but CLIA fields with no CLIA.  Will open a new issue to deal with selecting the proper fields for CLIA

### CODE REVIEW

**Review by jimduff-usds — COMMENTED** [review]
@carlosfelix2, is this PR meant to fix the problem that the `equipment_model_name` is not getting filled in?   If so, it doesn't work for me.       How I'm testing for this:
1.  Using this branch.
2. In Hl7Serializer.kt, line 318, I'm setting the logger to `logger.info,` so I can see detailed output.
2. Then I run the following
`./prime data --input-schema hl7/hl7-ingest-covid-19 --input /<yourpathgoeshere>/care_evolution_0415.hl7 --route --output-dir ./junk | grep equipment_model_name`

And I get an empty value:
`14:12:08.612 [main][1] INFO  gov.cdc.prime.router.serializers.Hl7Serializer - equipment_model_name -> `

Confusingly, there is no warning that the field is empty, but equipment_model_name is empty, so the test doesn't pass the 'qualityFilter'.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 1]
```diff
@@ -183,16 +180,48 @@ class Hl7Serializer(val metadata: Metadata): Logging {
...
+
+        /**
+         * Decode answers to AOE questions
+         * @param element the element for the AOE question
+         * @param terser the HAPI terser
+         * @param errors the list of errors for this message decoding
+         * @param warnings the list of warnings for this message decoding
+         * @return the value from the HL7 message or an empty string if no value found
+         */
+        fun decodeAOEQuestion(element: Element, terser: Terser, errors: MutableList<String>,
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:194]: > I'm guessing this is pretty fragile code - we'll see the AOE questions showing up in all kinds of inventive formats.

That is possible.  I have not touched this logic, so we could create a new issue if we want some specific functionality out of this.

**Comment by carlosfelix2** [comment]
> @carlosfelix2, is this PR meant to fix the problem that the `equipment_model_name` is not getting filled in? If so, it doesn't work for me. How I'm testing for this:
> 
> 1. Using this branch.
> 2. In Hl7Serializer.kt, line 318, I'm setting the logger to `logger.info,` so I can see detailed output.
> 3. Then I run the following
>    `./prime data --input-schema hl7/hl7-ingest-covid-19 --input /<yourpathgoeshere>/care_evolution_0415.hl7 --route --output-dir ./junk | grep equipment_model_name`
> 
> And I get an empty value:
> `14:12:08.612 [main][1] INFO gov.cdc.prime.router.serializers.Hl7Serializer - equipment_model_name -> `
> 
> Confusingly, there is no warning that the field is empty, but equipment_model_name is empty, so the test doesn't pass the 'qualityFilter'.

The main problem I had is that my new schema was not using the latest LIVD table.  I also adjusted a could of things for the device ID.  Try it again.

**Review by jimduff-usds — APPROVED** [review]
Yes, I can generate end to end - hl7 in, csv and hl7 out!

And I've done a do-no-harm test.    @mauricereeves , do you want to review the HL7 details?  I'd *love* to get this into our Thursday morning deploy.

**Review by MauriceReeves-usds — CHANGES_REQUESTED** [review]
[sample_hl7_message.hl7.txt](https://github.com/CDCgov/prime-data-hub/files/6554340/sample_hl7_message.hl7.txt)
[hl7-ingest-covid-19-83d7dd66-a0a1-481d-81bc-73be74039237-20210527131305.hl7.txt](https://github.com/CDCgov/prime-data-hub/files/6554344/hl7-ingest-covid-19-83d7dd66-a0a1-481d-81bc-73be74039237-20210527131305.hl7.txt)
[covid-19-ffaad964-fe30-4aca-b175-d20561d63994-20210512210501.hl7.txt](https://github.com/CDCgov/prime-data-hub/files/6554349/covid-19-ffaad964-fe30-4aca-b175-d20561d63994-20210512210501.hl7.txt)


There are a couple of issues here. The biggest is the fact that by removing all of the HL7 fields from the `covid-19` schema, HL7 serialization no longer works out of the box for anyone. The default schema name for HL7 generation is `covid-19`, and that means that when it attempts to serialize the a `Report` row you will get a skeleton HL7 file with no data in it.

The setting of this default schema is in `TranslatorConfiguration.kt` which was not changed to your hl7-ingest-covid-19 schema. That would be also cause problems. In addition, the field for `Hl7Configuration` is read-only so even if we updated every organization in the organization schema files to your new schema the fix would not work.

Also, while checking this, I manually updated the default schema for HL7 serialization to your ingestion schema and then ran a CSV file through Docker to compare what it generated, and there are obvious differences that we can't pass along to our clients. Whatever work you do for HL7 clean up must not affect the output that states are expecting.

Last note, pointing everything to your new schema also changes the name of the files that are output because the schema name is included in the file output by default, which is another change we would need to communicate to our clients.

For all of the reasons above, I can't approve this change.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 2]
```diff
@@ -202,65 +231,59 @@ class Hl7Serializer(val metadata: Metadata): Logging {
...
+                if (it.mapperRef != null) {
+                    // This gets the requiredvalue names, then gets the value from mappedRows that has the data
+                    val args = it.mapperArgs ?: emptyList()
+                    val valueNames = it.mapperRef.valueNames(it, args)
+                    val valuesForMapper = valueNames.mapNotNull { elementName ->
+                        val valueElement = schema.findElement(elementName)
+                        if(valueElement != null && mappedRows.containsKey(elementName) && !mappedRows[elementName].isNullOrEmpty()) {
+                            ElementAndValue(valueElement, mappedRows[elementName]!!.first())
+                        } else {
+                            null
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:271]: Hmm. The mapper algorithm should key on if a value is present and if a mapper is defined. The algorithm is: input_value ?: mapper value ?: default value.

**Review by RickHawesUSDS — CHANGES_REQUESTED** [review]
I think I found an error in the mapper logic. The trouble is that I've implemented this algorithm in two places (CSVSerializer and Translator) and now you've implemented it in a different place. My fault. We should have a common piece of code for this important logic. Can I work with you to make a common place for this code?

File: prime-router/src/main/kotlin/Mappers.kt [hunk 3]
```diff
@@ -605,6 +613,25 @@ class ZipCodeToCountyMapper : Mapper {
     }
 }
 
+/**
+ * This mapper performs no operation and is meant to override mappers set on parent schemas, so no mapper runs.
+ * Arguments: None
+ * Returns: null
+ */
+class NoMapper : Mapper {
+    override val name = "none"
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/Mappers.kt:622]: > I like this. Maybe NullMapper would be a better name? Not sure.

I like NullMapper.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt (hunk 2 above)
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:288]: > Hmm. The mapper algorithm should key on if a value is present and if a mapper is defined. The algorithm is: input_value ?: mapper value ?: default value.

I can change that and look into making this common code.  However, I am wondering if we would want to have a way to post-process the data, like split it or validate it.

**Comment by carlosfelix2** [comment]
> [sample_hl7_message.hl7.txt](https://github.com/CDCgov/prime-data-hub/files/6554340/sample_hl7_message.hl7.txt)
> [hl7-ingest-covid-19-83d7dd66-a0a1-481d-81bc-73be74039237-20210527131305.hl7.txt](https://github.com/CDCgov/prime-data-hub/files/6554344/hl7-ingest-covid-19-83d7dd66-a0a1-481d-81bc-73be74039237-20210527131305.hl7.txt)
> [covid-19-ffaad964-fe30-4aca-b175-d20561d63994-20210512210501.hl7.txt](https://github.com/CDCgov/prime-data-hub/files/6554349/covid-19-ffaad964-fe30-4aca-b175-d20561d63994-20210512210501.hl7.txt)
> 
> There are a couple of issues here. The biggest is the fact that by removing all of the HL7 fields from the `covid-19` schema, HL7 serialization no longer works out of the box for anyone. The default schema name for HL7 generation is `covid-19`, and that means that when it attempts to serialize the a `Report` row you will get a skeleton HL7 file with no data in it.
> 
> The setting of this default schema is in `TranslatorConfiguration.kt` which was not changed to your hl7-ingest-covid-19 schema. That would be also cause problems. In addition, the field for `Hl7Configuration` is read-only so even if we updated every organization in the organization schema files to your new schema the fix would not work.
> 
> Also, while checking this, I manually updated the default schema for HL7 serialization to your ingestion schema and then ran a CSV file through Docker to compare what it generated, and there are obvious differences that we can't pass along to our clients. Whatever work you do for HL7 clean up must not affect the output that states are expecting.
> 
> Last note, pointing everything to your new schema also changes the name of the files that are output because the schema name is included in the file output by default, which is another change we would need to communicate to our clients.
> 
> For all of the reasons above, I can't approve this change.

Restored the covid-19 schema.

File: prime-router/metadata/schemas/HL7/hl7-ingest-covid-19.schema [hunk 4]
```diff
@@ -0,0 +1,64 @@
+---
+name: hl7-ingest-covid-19
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/metadata/schemas/HL7/hl7-ingest-covid-19.schema:2]: > By the name of ingest, I assume that this is an ingest only schema?

Correct

File: prime-router/src/main/kotlin/LookupTable.kt [hunk 5]
```diff
@@ -122,7 +122,7 @@ class LookupTable(
      * Get the set of distinct values in a column in the table.
      */
     fun getDistinctValuesInColumn(selectColumn: String): Set<String> {
-        val selectColumnNumber = headerIndex[selectColumn.toLowerCase()] ?: return emptySet()
+        val selectColumnNumber = headerIndex[selectColumn.lowercase()] ?: return emptySet()
```
Comment (pablocaviglia) [inline prime-router/src/main/kotlin/LookupTable.kt:125]: Do you know why my IDE (intellij) doesn't accept the method 'lowercase'? It's weird that the one it accepts is the one you replaced.

**Review by RickHawesUSDS — APPROVED** [review]

File: prime-router/src/main/kotlin/LookupTable.kt (hunk 5 above)
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/LookupTable.kt:125]: > Do you know why my IDE (intellij) doesn't accept the method 'lowercase'? It's weird that the one it accepts is the one you replaced.

That happened to me.  You need to make sure to install the Kotlin 1.5.0 plugin, which may still be in beta or unnoficial.

File: prime-router/metadata/schemas/HL7/hl7-ingest-covid-19.schema [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+elements:
+  - name: device_id
+    type: TEXT
+    mapper: none()
+    hl7Field: OBX-17-1
+    hl7OutputFields: [ OBX-17-9 ]
+    documentation: Device_id is a generated value for the OBX-17 field. It is based on the device model and the LIVD table.
+
+  - name: device_id_type
+    type: ID
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/metadata/schemas/HL7/hl7-ingest-covid-19.schema:15]: Per our conversation, we will reset the types to the original covid 19 schema types

File: prime-router/metadata/schemas/HL7/hl7-ingest-covid-19.schema [hunk 7]
```diff
@@ -0,0 +1,64 @@
+---
+name: hl7-ingest-covid-19
+topic: covid-19
+description: Schema for ingesting HL7 messages
+extends: covid-19
+elements:
+  - name: device_id
+    type: TEXT
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/metadata/schemas/HL7/hl7-ingest-covid-19.schema:8]: Per our conversation, we will reset the types to the original covid 19 schema types

**Review by MauriceReeves-usds — APPROVED** [review]

**Pull request merged**